presto-main/src/main/java/com/facebook/presto/dispatcher/LocalDispatchQuery.java

```diff
@@ -94,6 +104,63 @@ public LocalDispatchQuery(
         });
     }
 
+    @Override
+    public void startWaitingForPrerequisites()
+    {
+        // It's possible that queryExecution fails before we start for prerequisites, in that case, don't even
+        // start waiting for prerequisites
+        if (isDone()) {
+            return;
+        }
+
+        try {
+            Session session = stateMachine.getSession();
+            CompletableFuture<?> prerequisitesFuture = queryPrerequisites.waitForPrerequisites(
+                    stateMachine.getQueryId(),
+                    new QueryPrerequisitesContext(
+                            session.getCatalog(),
+                            session.getSchema(),
+                            stateMachine.getBasicQueryInfo(Optional.empty()).getQuery(),
+                            session.getSystemProperties(),
+                            session.getConnectorProperties()));
+
+            addStateChangeListener(state -> {
+                if (state.isDone()) {
+                    queryPrerequisites.queryFinished(stateMachine.getQueryId());
+                    if (!prerequisitesFuture.isDone()) {
+                        prerequisitesFuture.cancel(true);
+                    }
+                }
+            });
+
+            prerequisitesFuture.whenCompleteAsync((result, throwable) -> {
```

Comment:
My understanding is that if I don't use the async function - then depending on when the future finishes, that will decide the execution - 

1. If the future is finished before the callback is set - this will execute in the thread where the callback is registered (which is fine for us)
2. If the future is finished after the callback is set - this will execute in the thread that completes the future - which is not desirable in my opinion - we should not load the threads that are computing the prerequisites with actual query registration stuff.

So given (2) - I went with the async option

```
@Test
    public void testFutures()
            throws Exception
    {
        CompletableFuture<?> a = CompletableFuture.completedFuture(null);
        System.out.println(1);
        a.whenComplete((x, y) -> {System.out.println(Thread.currentThread().getName());});
        System.out.println(2);
        Thread foo = new Thread(() -> a.whenComplete((x, y) -> {System.out.println(Thread.currentThread().getName());}), "test-thread");
        foo.start();
        foo.join();

        CompletableFuture<?> b = new CompletableFuture<>();
        System.out.println(3);
        b.whenComplete((x, y) -> {System.out.println(Thread.currentThread().getName());});
        System.out.println(4);
        Thread bar = new Thread(() -> b.complete(null), "test-thread2");
        bar.start();
        bar.join();
```

Output:

```
1
main
2
test-thread
3
4
test-thread2
```